### ISSUE TITLE
PHP Fatal error:  Call to member function loadTranslations() on null

### ISSUE THREAD

**Issue by ReinVelt** [simplesamlphp/simplesamlphp#1592]
Got an error message at my IDP:

PHP Fatal error: Uncaught Error: Call to member function loadTranslations() on null in /var/simplesamlphp/lib/SimpleSAML/Locale/Localization.php:261

This issue is also mentioned in https://github.com/simplesamlphp/simplesamlphp/issues/905 but that one is closed

I using the latest current release 1.19.5

**Comment by tvdijen** [comment]
Thanks @ReinVelt ! We'll look into it

**Comment by tvdijen** [comment]
Just to be clear, you're not using the 'newui' option, are you?

**Comment by ReinVelt** [comment]
i don't know. How can i see that?  
i did not specify a usenewui option anywhere so probably not using it.

**Comment by tvdijen** [comment]
It's a hidden setting, so you're not using it then :) Thanks!

**Comment by tvdijen** [comment]
I really can't wrap my head around this.. Were you upgrading a previous installation?

**Comment by ReinVelt** [comment]
I cannot reproduce this any more.. 
We were adding a new idp so maybe it is caused by a configuration error.

**Comment by pacham7** [comment]
This solution worked for me:
```
I found the answer, the location it was not initializing. It was a bug. In the ./lib/SimpleSAML/Locale/Localization.php file I only call this method:

$this->setupTranslator();

in the end of the constructor and it works correctly.
```

https://stackoverflow.com/questions/52955851/simplesamlphp-missing-some-configuration-what-is-missing

**Comment by tvdijen** [comment]
That only makes sense if you have set `language.i18n.backend` to `gettext/gettext`, while `usenewui` is set to `false`...

**Comment by areis422** [comment]
I'm also running into this after an upgrade from 1.19.0 to 1.19.5.

Only seems to occur when clicking on the Lookup button on the Federation page

simplesaml/module.php/core/show_metadata.php?set=saml20-idp-remote&entityid=

assuming because entityid is empty....

**Comment by tvdijen** [comment]
@areis422 Thanks for the additional info! This is something we can work with!

**Comment by thijskh** [comment]
It's specifically Error pages that are broken. It was broken by 8215060c63916cf6c64fc84fded90882fa73f432 which only works with the usenewui setting turned on. And further broken by the yet-unreleased 23278dccf77277cb7c12b0393bac3940cdc33207 which also only works with usenewui on.

**Comment by thijskh** [comment]
Maybe I'm thinking too simple, but... maybe just revert those two commits for 1.19, spend no more time on trying to fix usenewui in 1.19, instead just release 2.0?

**Comment by tvdijen** [comment]
I agree marking this as a WONTFIX...   I honestly wouldn't know how to fix this particular issue for the old ui,

**Comment by thijskh** [comment]
So we'd have to back out those commits. Because breaking the error pages for oldui (the default) is I think not good to have.

**Comment by jad-alcantara** [comment]
https://github.com/simplesamlphp/simplesamlphp/blob/8215060c63916cf6c64fc84fded90882fa73f432/lib/SimpleSAML/Locale/Localization.php#L258

I added 'LC_MESSAGES' inbetween the $langPath and $poFile variables and was able to get it to work.  In my case, it was trying to retrieve the core.php file, but it was missing another directory before it could reach it.  

Not sure if the other modules are structured the same way, so I wouldn't say if this would be the exact fix.

**Comment by tvdijen** [comment]
That doesn't make much sense....  How van $langPath be without that?

**Comment by jad-alcantara** [comment]
AH Disregard my last comment?  I'm honestly not sure why, but it started working without that added path I mentioned and the issue I had with getting the locale for core never came back up 🤔  

So sorry for the confusion!  🙇🏼

**Comment by tvdijen** [comment]
The confusion gets even worse.. Please, get me a diff of what you have changed to make it work..

**Comment by cicnavi** [comment]
Hi, 
we have ran into the same error when testing v1.19.5 (we moved from v1.18.8). We have encountered this when we tried SAML authn from an SP which was not registered in metadata/saml20-sp-remote.php, which triggered 'Unable to locate metadata for 'non-registered-entity-id'" error. 

From my understanding the problem is in https://github.com/simplesamlphp/simplesamlphp/blob/aecc7bbaed9c8f846b300d6076118c49b22e637a/lib/SimpleSAML/Locale/Localization.php#L259

```
if (file_exists($poPath) && is_readable($poPath)) {
      $translations = Translations::fromPoFile($poPath);
      $this->translator->loadTranslations($translations);
  } else {
      $error = "Localization file '$poFile' not found in '$langPath', falling back to default";
      Logger::debug($_SERVER['PHP_SELF'] . ' - ' . $error);
  }
```

So, the 'if' statement checks if 'po' files exist, and if so, it expects that the $this->translator is initialized. However, $this->translator is not always available, as per 

https://github.com/simplesamlphp/simplesamlphp/blob/aecc7bbaed9c8f846b300d6076118c49b22e637a/lib/SimpleSAML/Locale/Localization.php#L289

```
private function setupL10N(): void
  {
      if ($this->i18nBackend === self::SSP_I18N_BACKEND) {
          Logger::debug("Localization: using old system");
          return;
      }

      $this->setupTranslator();
      // setup default domain
      $this->addDomain($this->localeDir, self::DEFAULT_DOMAIN);
}
```

So, the $this->setupTranslator(); (which initializes $this->translator) only gets called if $this->i18nBackend is not self::SSP_I18N_BACKEND.

Our quick fix to test things out was to check for $this->translator together with 'po' files check in lib/SimpleSAML/Locale/Localization.php:259...

```
if (file_exists($poPath) && is_readable($poPath) && $this->translator) {...}
```

**Comment by tvdijen** [comment]
Hi @cicnavi ! Thanks, but I don't see how that fixes anything.. You would fall back on English translations completely instead of falling back to the old localization-system, right?

**Comment by thijskh** [comment]
https://github.com/simplesamlphp/simplesamlphp/issues/1592#issuecomment-1072298875

This is the solution I think.

**Comment by michielbdejong** [comment]
Thanks for working on a fix! As an end-user who just installed v1.19.5 from scratch, I was able to work around it with @pacham7 's suggestion from https://github.com/simplesamlphp/simplesamlphp/issues/1592#issuecomment-1059107819 - maybe you can add this in a 'Known Issues' section in your install instructions somewhere?

**Comment by Tequenikality** [comment]
> This solution worked for me:
> 
> ```
> I found the answer, the location it was not initializing. It was a bug. In the ./lib/SimpleSAML/Locale/Localization.php file I only call this method:
> 
> $this->setupTranslator();
> 
> in the end of the constructor and it works correctly.
> ```
> 
> https://stackoverflow.com/questions/52955851/simplesamlphp-missing-some-configuration-what-is-missing

This solution worked perfectly for me - thanks! It was really confusing for me, because I run a Debian 11 server and I had to re-image it because I was switching from a Mirrored raid array to a Parity one. I copied all config files exactly, and even downloaded a list of the installed packages. Everything was as close to 100% copied as I could feasibly get - it took five hours to copy everything over, but the new installation just wasn't working. After looking through the PHP log, I found the error and after some research - this solution.

**Comment by thijskh** [comment]
I have backed out the two mentioned commits for the 1.19 branch. This should restore error messages in the default configuration, at the cost of a fix for those running the experimental `usenewui`. We can fix that properly in 2.0.

Note that if you ran into this problem, SimpleSAMLphp was trying to display an error page so likely you still have a further issue to solve.

**Comment by vahem2lu** [comment]
In my case "issue" was that one SP metadata was missing (which would be a okay error to display). :smile:

**Comment by santosalan** [comment]
> This solution worked for me:
> 
> ```
> I found the answer, the location it was not initializing. It was a bug. In the ./lib/SimpleSAML/Locale/Localization.php file I only call this method:
> 
> $this->setupTranslator();
> 
> in the end of the constructor and it works correctly.
> ```
> 
> https://stackoverflow.com/questions/52955851/simplesamlphp-missing-some-configuration-what-is-missing

This solution worked perfectly for me - thanks!

**Comment by tvdijen** [comment]
The fix was added in the 1.19.6 release, so we can close this issue
